### PR TITLE
driver: i2c: TCA9546a: Add support for more I2C mux

### PR DESCRIPTION
**Pull request by aaillet**
Hi, 
Initial goal was to support the the PCA9548 8 channel mux I'm working with.
As discussed with @kurddt , this PR is containing modification in order for the I2C mux driver to :
- Support more mux from the TCA954x and PCA954x families.
- Support any mux size that could exist in these families.
- Support reset signals for mux that are equipped with a reset input pin.

This driver has been tested with an 8-channel with reset PCA9548A mux from TI and should not break compatibility with previously tested mux.

### CODE REVIEW

**Review by kurddt — CHANGES_REQUESTED** [review]
I don't know what is the policy about changing existing driver's compatible property. @mbolivar could you advice on the matter ?

**Review by kurddt — CHANGES_REQUESTED** [review]

**Review by julien-massot — APPROVED** [review]

**Review by kurddt — APPROVED** [review]

**Comment by aaillet** [comment]
Rebased on top of main !

**Comment by kurddt** [comment]
After some looking around in the zephyr source tree and thoughts, I think it would be better from the driver user's point of view to handle the different variant of the i2c mux like what is done for the mcp320X or lmp90xxx adc drivers:

A binding file with different compatible exists for each supported variant. That way the `nchans` devicetree property is not needed but deduced from the compatible. 

That said I won't block the PR for that, if you don't want to make that change I will do this work after the PR is merged

**Comment by aaillet** [comment]
I don't know if it's the right solution to multiply the number of binding (one per reference) especially since they are for the most part, empty.
But if @mbolivar confirms that this is the solution applied in our case, we might as well not enter this PR and do it directly to avoid breaking the "compatible" several times.
In this case, I could do it myself based on the drivers you pointed out.

**Comment by mbolivar-nordic** [comment]
Changing compatibles is not ideal, but we do do it sometimes if we have to.

Can you please update the release notes to reflect what happened and how users can recover, though? Here is an example from 2.7:

https://docs.zephyrproject.org/latest/releases/release-notes-2.7.html#build-and-infrastructure

(See "Devicetree bindings" section in that link)

**Comment by kurddt** [comment]
@mbolivar-nordic and what are your thoughts on having several compatibles ? one for each supported device

**Comment by mbolivar-nordic** [comment]
@kurddt in general? I am in favor of it wherever it makes sense. Note that you can create multiple bindings that share properties using `include`, so e.g.

```
# my-common-binding.yaml
properties:
  foo:
     ...
  bar:
     ...

# vnd,device1.yaml
compatible: vnd,device1
include: my-common-binding.yaml

# vnd,device2.yaml
compatible: vnd,device2
include: my-common-binding.yaml
```

etc.

If that makes sense here, then I'm in favor of doing that.

```

**Comment by kurddt** [comment]
@aaillet Could you make the change ? and sorry to not have thought of that earlier

**Comment by aaillet** [comment]
> @aaillet Could you make the change ? and sorry to not have thought of that earlier

Hi, I'm gonna do the changes soon.

**Comment by aaillet** [comment]
Hi, just pushed new version of the driver.
The driver is now managing multiple compatible fields (tca9546a and tca9548a at the moment).
Each device version binding file is now declaring its own device compatible field.
Each device binding is also declaring its own child binding compatible to ensure backward compatibility and simple dts fields completion for users.
The driver has been tested using a pca9548a from NXP and everything is working as expected.
if I forgot to mention something, don't hesitate.
Aymeric

**Comment by aaillet** [comment]
Please @mbolivar-nordic , tell me if this bindings/device tree implementation is acceptable. Thanks

**Comment by kaolpr** [comment]
I've tried this PR on TCA9548A and can confirm that it works. However, I had to change `I2C_TCA954X_CHANNEL_INIT_PRIO` to be of value greater than `I2C_TCA954X_CHANNEL_ROOT_PRIO` to avoid not ready error (to initialize root prior to channels).

**Comment by aaillet** [comment]
> I've tried this PR on TCA9548A and can confirm that it works. However, I had to change `I2C_TCA954X_CHANNEL_INIT_PRIO` to be of value greater than `I2C_TCA954X_CHANNEL_ROOT_PRIO` to avoid not ready error (to initialize root prior to channels).

Hi, thank you for testing.
When testing on my board, I also have to set `I2C_TCA954X_CHANNEL_INIT_PRIO` to a greater value than `I2C_TCA954X_CHANNEL_ROOT_PRIO` following an advice that @kurddt gave [this issue](https://github.com/zephyrproject-rtos/zephyr/issues/37786#issuecomment-901862285).
Maybe we should set these by default for any board, but I let @kurddt tell us if he thinks it's a good idea !

**Review by kurddt — CHANGES_REQUESTED** [review]

**Comment by kurddt** [comment]
> > I've tried this PR on TCA9548A and can confirm that it works. However, I had to change `I2C_TCA954X_CHANNEL_INIT_PRIO` to be of value greater than `I2C_TCA954X_CHANNEL_ROOT_PRIO` to avoid not ready error (to initialize root prior to channels).
> 
> Hi, thank you for testing. When testing on my board, I also have to set `I2C_TCA954X_CHANNEL_INIT_PRIO` to a greater value than `I2C_TCA954X_CHANNEL_ROOT_PRIO` following an advice that @kurddt gave in [this issue](https://github.com/zephyrproject-rtos/zephyr/issues/37786#issuecomment-901862285). Maybe we should set these by default for any board, but I let @kurddt tell us if he thinks it's a good idea !

The best solution would be that I2C_TCA954X_CHANNEL_ROOT_PRIO = I2C_TCA954X_CHANNEL_INIT_PRIO + 1. But it's neither possible to do so when defining the device nor in KConfig files. The other solution would be to have the default set to arbitrary chosen value but I suspect this will probably also break for a lot of case. Also the `I2C_TCA954X_CHANNEL_ROOT_PRIO` need to be higher than the i2c bus controller driver init priority the TCA954x is connected to.

I will open an issue to add a compilation error when I2C_TCA954X_CHANNEL_INIT_PRIO <= I2C_TCA954X_CHANNEL_ROOT_PRIO. So that users are aware to set those according to their usage

File: drivers/i2c/i2c_tca954x.c
Comment (aaillet) [reply drivers/i2c/i2c_tca954x.c]: forgot to do it when switching to "gpio_dt_spec", just fixed and tested it, thanks

**Review by kurddt — APPROVED** [review]

**Review by julien-massot — APPROVED** [review]

File: drivers/i2c/i2c_tca954x.c
Comment (mbolivar-nordic) [inline drivers/i2c/i2c_tca954x.c:195]: You are using the generated macros directly here; this is not allowed. These are an implementation detail that should not be used directly from a driver. Please use DT_INST_FOREACH_STATUS_OKAY instead.

File: drivers/i2c/Kconfig.tca954x
Comment (mbolivar-nordic) [inline drivers/i2c/Kconfig.tca954x:7]:
```suggestion
	  Enable TCA954x series I2C bus switch
```

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Generally the DT bits look fine; thank you. I did see one issue that is a blocker for me, but otherwise I think it's OK.

I noticed a minor typo while I was looking at the Kconfig.

File: drivers/i2c/i2c_tca954x.c
Comment (aaillet) [reply drivers/i2c/i2c_tca954x.c:195]: I think that replacing `TCA954X_FOREACH_STATUS_OKAY` by `DT_FOREACH_STATUS_OKAY` is the cleanest way to do it with DT macros.
Being based on the adc lmp90xxx driver, do you want me to push another PR to fix this in this one too ?

File: drivers/i2c/Kconfig.tca954x
Comment (aaillet) [reply drivers/i2c/Kconfig.tca954x:7]: will fix, thanks

File: drivers/i2c/i2c_tca954x.c
Comment (aaillet) [reply drivers/i2c/i2c_tca954x.c:195]: After testing, it wont work since `DT_FOREACH_STATUS_OKAY` is declaring `DT_N_S_soc_S_i2c_e6510000_S_pca9548_71_S_mux_i2c_2_REG_IDX_0_VAL_ADDRESS` while my calls for `DT_INST` are searching for `DT_N_INST_DT_N_S_soc_S_i2c_e6510000_S_pca9548_71_ti_tca9548a_REG_IDX_0_VAL_ADDRESS`.

So if `DT_INST_FOREACH_STATUS_OKAY` cannot take a compat as argument and there isn't an equivalent to `DT_INST` that is `UTIL_CAT(DT_N_, DT_DASH(inst, compat))` the way I did it is the only way I see to do it !

Comment (aaillet) [reply drivers/i2c/i2c_tca954x.c:195]: After some testing, it's not possible to use `DT_INST_FOREACH_STATUS_OKAY` because it wont take compat as argument and we need it as soon as we support multiple compatible.
I tried using `DT_FOREACH_STATUS_OKAY` but it wont work either because `DT_INST` is then searching for a `DT_INST_FOREACH_STATUS_OKAY` generated field.

That's why we use `TCA954X_FOREACH_STATUS_OKAY` to do `DT_INST_FOREACH_STATUS_OKAY` that can take a compat as argument.
I tried multiple ways to do it but can't find any that is working except the one already used.

Comment (mbolivar-nordic) [reply drivers/i2c/i2c_tca954x.c:195]: You can `#undef DT_DRV_COMPAT` and then define it to something else before using `DT_INST_FOREACH_STATUS_OKAY`. See, for example, drivers/serial/uart_gecko.c.

Comment (aaillet) [reply drivers/i2c/i2c_tca954x.c:195]: Just tested this :
```
/*
 * TCA9546A: 4 channels
 */
#define TCA9546A_INIT(n) TCA954x_ROOT_DEFINE(n, 4, ti_tca9546a)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_tca9546a
DT_INST_FOREACH_STATUS_OKAY(TCA9546A_INIT)

/*
 * TCA9548A: 8 channels
 */
#define TCA9548A_INIT(n) TCA954x_ROOT_DEFINE(n, 8, ti_tca9548a)
#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_tca9548a
DT_INST_FOREACH_STATUS_OKAY(TCA9548A_INIT)
```

It's working, is there a way to do it in a cleaner way or it is what you were thinking of @mbolivar-nordic ? Thanks

Comment (aaillet) [reply drivers/i2c/i2c_tca954x.c]: fixed it

**Comment by aaillet** [comment]
Replaced `TCA954X_FOREACH_STATUS_OKAY` by redefining `DT_DRV_COMPAT` for each device supported.
Also renamed config and data structure in order to support multiple device version on one and only board (for both root and channels structures).
Rebased on top of main.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**
